### PR TITLE
Stateless primary relocations

### PR DESCRIPTION
**Pull request by DaveCTurner**
Adds support for proper online relocations of stateless primaries.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/RoutingNodesTests.java [hunk 1]
```diff
@@ -484,18 +478,13 @@ public void testMoveShardWithPromotableOnlyRole() {
...
-            0L,
-            new RoutingChangesObserver() {
-            }
-        );
+        routingNodes.relocateShard(routingNodes.node("node-1").getByShardId(shardId), "node-3", 0L, new RoutingChangesObserver() {
+        });
 
-        assertThat(routingNodes.node("node-1").getByShardId(shardId), nullValue());
-        assertThat(routingNodes.node("node-2").getByShardId(shardId), nullValue());
+        assertThat(routingNodes.node("node-1").getByShardId(shardId).state(), equalTo(RELOCATING));
```
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/RoutingNodesTests.java:484]: This is the only part that I had not seen before in your changes. I see we're asserting an intermediate state here where the relocation has not fully completed yet (it's in transit from node-1 to node-3). Can we also wait for the completion of the relocation here and assert that it's finally only in node-3?

**Review by kingherc — APPROVED** [review]
I've seen most of this before, so LGTM. One comment on the test only.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/RoutingNodesTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/RoutingNodesTests.java:484]: Yeah we hadn't run these tests before, only the serverless ones, so I expect there might be a few changes needed.

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/part-1 (https://github.com/elastic/elasticsearch/issues/97183)
@elasticmachine please run elasticsearch-ci/part-3 (https://github.com/elastic/elasticsearch/issues/80703)

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/RoutingNodesTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/RoutingNodesTests.java:456]: There is no actual relocation in progress here, this is only to verify that the routing table changes as we expect. This test (and the one above) were added to support the temporary change that triggered stateless relocations by just closing all the shards, so now really this test can say that the routing table behaves the same for stateless and regular relocations. I've removed the duplication in d4ca54f.

**Pull request merged**